pkg/apis/application/v1alpha1/types.go

```diff
@@ -2341,3 +2375,66 @@ func (d *ApplicationDestination) MarshalJSON() ([]byte, error) {
 	}
 	return json.Marshal(&struct{ *Alias }{Alias: (*Alias)(dest)})
 }
+
+func (proj *AppProject) NormalizeJWTTokens() {
+	for i, role := range proj.Spec.Roles {
+		for j, token := range role.JWTTokens {
+			if token.ID == "" {
+				token.ID = strconv.FormatInt(token.IssuedAt, 10)
+				role.JWTTokens[j] = token
+			}
+		}
+		proj.Spec.Roles[i] = role
+	}
+
+	for _, roleTokenEntry := range proj.Status.JWTTokensByRole {
+		for j, token := range roleTokenEntry.Items {
+			if token.ID == "" {
+				token.ID = strconv.FormatInt(token.IssuedAt, 10)
+				roleTokenEntry.Items[j] = token
+			}
+		}
+	}
+
+	syncJWTTokenBetweenStatusAndSpec(proj)
+}
+
+func syncJWTTokenBetweenStatusAndSpec(proj *AppProject) {
+	for roleIndex, role := range proj.Spec.Roles {
+		tokensInSpec := role.JWTTokens
+		tokensInStatus := []JWTToken{}
+		if proj.Status.JWTTokensByRole == nil {
+			tokensByRole := make(map[string]JWTTokens)
+			proj.Status.JWTTokensByRole = tokensByRole
+		} else {
+			tokensInStatus = proj.Status.JWTTokensByRole[role.Name].Items
+		}
+		tokens := jWTTokensCombine(tokensInStatus, tokensInSpec)
+		proj.Spec.Roles[roleIndex].JWTTokens = tokens
+		proj.Status.JWTTokensByRole[role.Name] = JWTTokens{Items: tokens}
+	}
+}
+
+func jWTTokensCombine(tokens1 []JWTToken, tokens2 []JWTToken) []JWTToken {
```

Comment:
I think we should use map to avoid n^2 complexity:

```golang
func jWTTokensCombine(tokens1 []JWTToken, tokens2 []JWTToken) []JWTToken {
	combined := make(map[string]JWTToken)
	for i := range tokens1 {
		combined[tokens1[i].ID] = tokens1[i]
	}
	for i := range tokens2 {
		combined[tokens2[i].ID] = tokens2[i]
	}
	var tokens []JWTToken
	for k := range combined {
		tokens = append(tokens, combined[k])
	}

	return tokens
}
```